runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -5529,42 +5538,27 @@ MM_CopyForwardScheme::cleanCompressedSurvivorCardTable(MM_EnvironmentVLHGC *env)
 }
 
 void
-MM_CopyForwardScheme::abandonTLHRemainder(MM_EnvironmentVLHGC *env, bool preserveRemainders)
+MM_CopyForwardScheme::abandonTLHRemainders(MM_EnvironmentVLHGC *env, bool preserveRemainders)
 {
-	for (uintptr_t compactGroup = 0; compactGroup < _compactGroupMaxCount; compactGroup++) {
+	bool preserved = false;
+	uintptr_t tenureRemainderSize = 0;
+	for (intptr_t compactGroup = _compactGroupMaxCount - 1; compactGroup >= 0; compactGroup--) {
 		MM_CopyForwardCompactGroup *compactGroupForMarkData = &(env->_copyForwardCompactGroups[compactGroup]);
-		if (NULL != compactGroupForMarkData->_TLHRemainderBase) {
-			Assert_MM_true(NULL != compactGroupForMarkData->_TLHRemainderTop);
-			env->_cycleState->_activeSubSpace->abandonHeapChunk(compactGroupForMarkData->_TLHRemainderBase, compactGroupForMarkData->_TLHRemainderTop);
-			if (!preserveRemainders) {
-				/* discard the remainder for nursery regions */
-				discardHeapChunk(compactGroupForMarkData, compactGroupForMarkData->_TLHRemainderBase, compactGroupForMarkData->_TLHRemainderTop);
-				compactGroupForMarkData->resetTLHRemainder();
-			}
-		} else {
-			Assert_MM_true(NULL == compactGroupForMarkData->_TLHRemainderTop);
-		}
-	}
-}
 
-void
-MM_CopyForwardScheme::resetAllTLHRemainders(MM_EnvironmentVLHGC *env)
-{
-	MM_CopyForwardCompactGroup *copyForwardCompactGroups = NULL;
-	for (uintptr_t id = 0; id < _extensions->gcThreadCount; id++) {
-		copyForwardCompactGroups = &_compactGroupBlock[id * _compactGroupMaxCount];
+		preserved = !compactGroupForMarkData->abandonTLHRemainder(env, compactGroup, preserveRemainders);
+		uintptr_t age = MM_CompactGroupManager::getRegionAgeFromGroup(env, compactGroup);
```

Comment:
you could swap these two lines and pass age to abandon method (it does not really need exact compact group index)

Reply:
while it would slightly clean up abandon method, it would complicate the other call site of abandon (in clearCache), so probably not worth it

however, try to put the aging logic into a separate method
or piggyback to abandon method (instead of passing boolean preserveRemainders you can pass  a pointer to potential next age compact group or NULL)

Reply:
simply method to 	`MMINLINE bool abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t age = 0)`
 default age == 0 imply preserve == false.